Spatial_Engine: Add Query.IsFullyOnGrid() method

IsakLarbornBH opened the pull request:
<!-- PLEASE ENSURE YOU REVIEW THE CONTENT OF EACH PR CAREFULLY, INCLUDING SUBSEQUENT COMMENTS BY YOURSELF OR OTHERS. -->
<!-- IN PARTICULAR PLEASE ENSURE THAT SENSITIVE OR INAPPROPRIATE INFORMATION IS NOT UPLOADED -->


   
### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

Closes #2110 

Adds Query.IsFullyOnGrid() method to Spatial_Engine. This method accepts an IElement, a Level, and a double tolerance as input, and tests the IElement for proximity to the Grid, returning true if the IElement is fully contained within the set tolerance from the Grid curve.

<!-- Add short description of what has been fixed -->


### Test files
<!-- Link to test files to validate the proposed changes -->
[IsFullyOnGrid.zip](https://github.com/BHoM/BHoM_Engine/files/5651932/IsFullyOnGrid.zip)
(Still don't have my sharepoint access back, so the test file is here)

### Changelog
<!-- Text to go into changelog if applicable -->
<!-- Please see https://github.com/BHoM/documentation/wiki/changelog for guidelines -->
`Query.IsFullyOnGrid()` Query method added in the `Spatial_Engine` for `IElement` class

### Additional comments
<!-- As required -->
Currently the distance input is called `tolerance` and defaults to `BH.oM.Geometry.Tolerance.Distance` pending a final decision on #2158 IsFullyOnLevel. These two methods should have the same name and behaviour for these inputs before merging. (Update: IsFullyOnLevel is now merged, using tolerance in this way.)

This method does not properly support NurbsCurves, as it uses control points for evaluation, which works well for Lines and Arcs. To properly support NurbsCurves a more sophisticated algorithm capable of finding the curve's farthest point in distance from the Grid curve would be required. As it currently stands, this method throws a warning if it detects that its input is a NurbsCurve, but is not prevented from running.

FraserGreenroyd commented on Spatial_Engine/Query/IsFullyOnGrid.cs:
```diff
@@ -0,0 +1,132 @@
+                    return false;
+            }
+
+            return ctrlPts.Count > 0;
```
Won't this will always be true? And might thus be easier to just do `return true;`?

IsakLarbornBH replied:
You are probably right. It's just I've been working on null checks, and that made me wonder what would happen if you fed it an empty curve object with no curve segments. Since I didn't want it to return true in that case or some other when there are no control points returned, I added this check. But I don't think there's currently a way to trigger this that doesn't just crash in one of one of the earlier methods.

FraserGreenroyd reviewed (changes requested):
Minor comment but otherwise this looks good @IsakLarbornBH - we should probably have a catch up with @IsakNaslundBh next week to bottom out the points you're raising regarding the other PR as well to unblock you :smile:

IsakNaslundBh commented on Spatial_Engine/Query/IsFullyOnGrid.cs:
```diff
@@ -0,0 +1,132 @@
+        [Output("isFullyOnGrid", "A boolean which is true if the geometrical representation of an IElement1D is fully within a set tolerance from the grid line.")]
+        public static bool IsFullyOnGrid(this IElement1D element1D, Grid grid, double tolerance = BH.oM.Geometry.Tolerance.Distance)
+        {
+            ICurve curve = element1D.IGeometry().IProject(Plane.XY);
+            List <Point> ctrlPts = curve.ControlPoints();
```
Think we are better off projecting the controlpoints rather than the curve, as projecting the curve actually might turn it into a nurbs curve. (arcs and circles for example). Could help with your check below.

This is, ofc, unless you had it thew way you had it before intentionally

```suggestion
            ICurve curve = element1D.IGeometry();
            List <Point> ctrlPts = curve.ControlPoints().Select(x => x.Project(Plane.XY)).ToList();
```

Can also add, though, that checking the control points might not be fully rigorous (arc vs a polyline for example, might give you a false "true", but lets catch up on that next week.)

IsakLarbornBH replied:
That does sound like a good idea. There's no particular reason why the curve is projected rather than the points.

You're right about the potential for false positives. I initially had the idea of using gradient descent to find local maximi in distance, but I figured it might be too heavy and complex for a method like this. I'm sure there are many more ways of making better algorithms than what is there currently. Let's catch up on that.

IsakLarbornBH replied:
I got rid of this check.

IsakNaslundBh reviewed (commented):
Initial comment regarding the projection

IsakLarbornBH commented:
>Initial comment regarding the projection

Points, rather than curves, are now projected to the XY plane for checking. I have also disabled support for curved grids as discussed. It would be possible to reduce the likelihood of false positives for curved grids by doing an identical check for the grid against the projected curve, but a more sophisticated algorithm would be even better and be able to support NurbsCurves so if curved grid support becomes necessary that would be the way to go.

IsakLarbornBH commented:
I've now added a new test file based on the one I made for IsFullyOnGrid. I still don't have my Sharepoint access back, or even a response from IT a week after I contacted them, so the file is uploaded here.

Also changed the E1D method to use the E0D method for checking the distance of control points, rather than having its own identical comparison.

IsakNaslundBh commented:
FYI @IsakLarbornBH I have put your testfile on sharepoint here:

https://burohappold.sharepoint.com/:f:/s/BHoM/EjkDVhX6IA9Gu2CLJHgHDBQBl6y3_XKQx28ed8NXIOwiFA?e=6KZ5vh

IsakNaslundBh commented:
@BHoMBot check compliance

IsakLarbornBH commented:
> FYI @IsakLarbornBH I have put your testfile on sharepoint here:

Thank you :)

I've done some tests of a few other algorithms but not really found anything that works better than this, at least not without being slow. Therefore I think it's better that this gets merged as is, so I've removed WIP status.

IsakNaslundBh reviewed (approved):
Think this looks good now and seem to be working fine. Approved!

IsakNaslundBh commented:
/azp run BHoM_Engine.CheckInstaller

IsakNaslundBh commented:
@FraserGreenroyd think your comment has been resolved, if you want to have another look?

FraserGreenroyd reviewed (approved):
My comment has been addressed, I'm happy, thanks @IsakNaslundBh and @IsakLarbornBH
